File: backend/src/models/Prediction.ts
import mongoose, { Document, Schema } from "mongoose";

export interface IExplanation {
  feature: string;
  value: any;
  contribution: string;
  severity: "INFO" | "WARNING" | "CRITICAL";
}

export interface IPrediction extends Document {
  userId?: mongoose.Types.ObjectId;
  url: string;
  prediction: "phishing" | "legitimate";
  isPhishing: boolean;
  probability: number;
  riskScore: number;
  riskLevel: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
  modelVersion: string;
  modelName: string;
  inferenceLatencyMs: number;
  featureSummary: Record<string, any>;
  explanations: IExplanation[];
  createdAt: Date;
}

const ExplanationSchema = new Schema<IExplanation>(
  {
    feature: { type: String, required: true },
    value: { type: Schema.Types.Mixed },
    contribution: { type: String, required: true },
    severity: { type: String, enum: ["INFO", "WARNING", "CRITICAL"], default: "INFO" }
  },
  { _id: false }
);

const PredictionSchema = new Schema<IPrediction>(
  {
    userId: { type: Schema.Types.ObjectId, ref: "User", index: true, default: null },
    url: { type: String, required: true, trim: true, maxlength: 2048 },
    prediction: { type: String, enum: ["phishing", "legitimate"], required: true, index: true },
    isPhishing: { type: Boolean, required: true, index: true },
    probability: { type: Number, required: true },
    riskScore: { type: Number, required: true },
    riskLevel: { type: String, enum: ["LOW", "MEDIUM", "HIGH", "CRITICAL"], required: true, index: true },
    modelVersion: { type: String, default: "1.0.0" },
    modelName: { type: String, default: "Phishing Hybrid Classifier" },
    inferenceLatencyMs: { type: Number, default: 0 },
    featureSummary: { type: Map, of: Schema.Types.Mixed, default: {} },
    explanations: [ExplanationSchema]
  },
  {
    timestamps: { createdAt: true, updatedAt: false }
  }
);

PredictionSchema.index({ createdAt: -1 });

export const Prediction = mongoose.model<IPrediction>("Prediction", PredictionSchema);

File: ml-service/app/schemas/prediction.py
"""
Pydantic Schemas for Request / Response Validation in FastAPI ML Microservice.
"""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class PredictRequest(BaseModel):
    url: str = Field(..., description="The full URL string to analyze for phishing risks", json_schema_extra={"example": "https://secure-login.appleid-verify.xyz/auth"})


class BatchPredictRequest(BaseModel):
    urls: List[str] = Field(..., description="List of URL strings to analyze", max_length=500)


class ExplanationItem(BaseModel):
    feature: str = Field(..., description="Name of the feature")
    value: Any = Field(..., description="Extracted value for this URL")
    contribution: str = Field(..., description="Human-readable explanation of how this contributed to the prediction")
    severity: str = Field("INFO", description="Severity level: INFO, WARNING, or CRITICAL")


class PredictResponse(BaseModel):
    url: str
    prediction: str = Field(..., description="'legitimate' or 'phishing'")
    is_phishing: bool
    probability: float = Field(..., description="Model probability of being phishing (0.0 to 1.0)")
    risk_score: float = Field(..., description="Risk score percentage (0 to 100)")
    risk_level: str = Field(..., description="Risk classification: LOW, MEDIUM, HIGH, CRITICAL")
    model_version: str
    model_name: str
    inference_latency_ms: float
    features: Dict[str, Any]
    explanations: List[ExplanationItem]
    disclaimer: str = "This prediction is an ML-based risk assessment and does not constitute 100% certainty."


class BatchItemResponse(BaseModel):
    url: str
    prediction: str
    is_phishing: bool
    probability: float
    risk_level: str
    risk_score: float


class BatchPredictResponse(BaseModel):
    total_scanned: int
    phishing_count: int
    legitimate_count: int
    total_time_ms: float
    results: List[BatchItemResponse]


class HealthResponse(BaseModel):
    status: str
    service: str
    version: str
    model_loaded: bool
    model_version: Optional[str]
    uptime_seconds: float


class FeatureMetadata(BaseModel):
    name: str
    category: str
    description: str


class FeaturesResponse(BaseModel):
    total_features: int
    features: List[FeatureMetadata]
